model/ftn/ww3_strt.ftn

```diff
@@ -193,6 +193,7 @@
       USE W3WDATMD
       USE W3ODATMD, ONLY: NDSE, NDST, NDSO, NAPROC, IAPROC,           &
                           NAPOUT, NAPERR, FNMPRE
+!/WRST      USE W3IDATMD, ONLY: W3NINP
```

Comment:
The WRST is used also for ESMF, and overwrites the forcing at the time step of the restart. It is an option for a specific application, not related to the spectra update. 

I consider that these 2 options should not be activated with the same switch, please update.

Reply:
Thanks Stelios

The use of WRST here is deliberate - when running w3uprstr with the UPD5/6 methods we need to read in the background wind speed along with the background wave spectrum. The best way to do this is via the restart file, rather than reading wind in from a separate output file. 

As far as I could tell WRST is the switch that controls wind being in the restart at all, so should be used identically in w3uprstr?? Otherwise I would need to set up the model with two switches, one (WRST) to tell me to have wind in the restart, and a second to tell w3uprstr to read wind from the restart. That seems a bit excessive...

Please let me know if I've missed a point and that WRST also affects the wind in a subsequent cycle in a way that would mess up the usual wind input for a non-coupled model.

Reply:
@flampouris please reply to @ukmo-ansaulter at your earliest convenience. I would like to have this PR merged to the WW3 develop as soon as possible. Thanks!

Reply:
Hello Andy @ukmo-ansaulter 

With activated the WRST, and for a specific combination of switches, the wind forcing at the time of the restart file is replaced by the wind information included in the restart file, see lines 2245-2255 at wmesmfmd.ftn.

In this case, the **analysis** of the wind field is replaced by a **prediction** of the wind field from an earlier cycle, resulting in discontinuity and potential degradation of the wind forcing.

Therefore, the choices for the update of the wavefield's initial conditions should not have any effect on the forcing.

I hope that this comment clarifies my original review. Please, let me know if there are any additional questions or concerns.

Reply:
@ukmo-ansaulter check also the PR #146

Reply:
I see exactly where you are - so WRST implies the actions of both storing wind data in the restart file and giving this wind data precedence over any new wind that uses the same timestamp. In order to get around this we would need to split WRST into two switches, one for each function and then use only the read/write wind in restart part in uprstr.

This doesn't sound like a great idea to me since you would then need to build your ESMF system with two switches to deal with the wind data. Also, from an uprstr point of view, the longer term development of UPD5/6 would want exactly this precedent for the wind to be in place - if we follow Lionello et al's method through completely, then the wind field in the restart should be updated in order to balance the updated wave field's wind-sea component.

As a work around, UPD5/6 already have an option in place where the background wind can be read from the same file as the analysed Hs field, so for those concerned about the precedence associated with WRST the wave model the code can be used in this way without the switch being activated.

Can I suggest therefore, that this is something that is made more explicit in the code and manual documentation rather than making the switch set up more complicated?

Reply:
@ukmo-ansaulter I agree. Could you add a runtime warning that informs the user of the effects of using a potentially unwanted effect when combining switches? @flampouris please comment.

Reply:
@ukmo-ansaulter

This could be a sufficient comprising solution, I suggest that reviewers from institutes using ESMF should approve the change.

My review is focused on the PR's code, we can continue the scientific discussion to a different forum.

Reply:
Thanks @flampouris - sounds good to me, we will get the changes implemented asap

Reply:
@ukmo-ansaulter and @ukmo-ccbunney thank you for your patience. Since requested, this review took 48 days to be initiated, I won't expect that you complete the required revisions in 48 hours, although it seems you have everything lined up to get this done shortly which I greatly appreciate. As a matter of respect to time and effort in our development process, I will discuss with @aliabdolali nd other trusted repo code managers ways to certify that our review process follows more reasonable timelines. Thank you.

Reply:
Noting that we took this conversation offline for a short period to clarify the risks and potential solutions for the change I am precis-ing below...

A check through the code indicates that building with the WRST switch carries no risk to the normal operation of the model; any contention is presently limited to the wind data used in coupled systems by ESMF.

The uprstr code, as stands, allows build and operation of the restart update without WRST being a mandatory switch; i.e. ESMF and DA can be operated safely in a consistent build with or without WRST

There is a minor risk of a user deciding to build  a model using the DA module and including WRST without realising the changed impact this would have on ESMF behaviour. There are two potential work arounds for this: 1) add a warning message to the ESMF module and reinforce in manual documentation; 2) update the switch used by ESMF to clobber the T+0 wind file

Based on the above, we have 3 courses of action dependent on preferred timescales and approach:

1. For an immediate merge: Accept PR as it is, including the minor risk that someone chooses to do something daft with the model running a develop branch in the short term. We then set up issues to work on either or both of work arounds 1 and 2 above

2. For a merge early next week: Met Office to update the code and PR with work around 1, assuming the PR would then be acceptable

3. Longer term: Reject the PR and split out the changes. PR with Chris' changes to be submitted late next week. The feature branch for DA to be re-submitted including work around 2) and with review by Jessica/ESMF developers at a later date TBD

This becomes a code managers prerogative now I think, but it is clearly important to get reviewers opinions here. @flampouris can you confirm if options 1 or 2 would be at all acceptable for you? @JessicaMeixner-NOAA could you also provide a view on the risk associated with 1 or 2 (so its not just my say so)?

Reply:
@ukmo-ansaulter thanks for summarizing the offline discussions, and thanks all for the great comments and comprehensive evaluation. Given the clearly-stated importance of final reviewer's opinions, which we fully agree, our prerogative is over-ridden and becomes conditional to having those opinions added to this comment thread. @flampouris and @jessica please add your views at your earliest convenience. Thanks!

Reply:
Thank-you @ukmo-ansaulter for providing such a helpful summary and thanks to all the reviewers for paying attention to such details.  As is understood by all, the WRST switch does not have any direct conflict with wmesmfmd or running of WW3 under ESMF.  The potential conflict is only that wmesmfmd:GetImport will overwrite the initial wind (wx0,wy0) with what is received from the coupling layer.  What is needed in wmesmfmd:GetImport is a (dynamic) flag to indicate that the initial winds are obtained from the restart and not to be overwritten.  Does such a flag exist in one of the w3*md modules?  If so, then we simply include it in wmesmfmd and add logic to wmesmfmd:GetImport accordingly.  This, I think, can be addressed after this merge has been completed.

Reply:
Thank you @tjcnrl for your assessment. Alongside other comments, arguments and evidence provided previously, this answers the only pending concern from @flampouris as folllows:

>     * This could be a sufficient comprising solution, I suggest that reviewers from institutes using ESMF should approve the change.

Issues have been created addressing other concerns that have been identified as enhancements and not essential modifications to this PR.

Once again, thanks to all reviewers for your comprehensive assessment of the proposed changes. I commend in particular reviews that were authentically constructive by going beyond simply identifying issues and risks, but also provided suggested solutions to such issues and risks in a way that fully leverages the strengths of the open development paradigm, towards making ours an improved software package.